openedx/core/djangoapps/user_api/preferences/api.py

```diff
@@ -0,0 +1,198 @@
+
+    Raises:
+        AccountUserNotFound: no user with username `username` exists (or `requesting_user.username` if
+            `username` is not specified)
+        AccountNotAuthorized: the requesting_user does not have access to change the account
+            associated with `username`
+        PreferenceUpdateError: the update could not be completed.
+    """
+    existing_user = _get_user(requesting_user, username)
+    try:
+        # First validate each preference setting
+        errors = {}
+        for preference_key in update.keys():
+            preference_value = update[preference_key]
+            try:
+                if preference_value is not None:
+                    UserPreference.validate_preference(existing_user, preference_key, preference_value)
+            except ValidationError as error:
+                errors[preference_key] = {
+                    "developer_message": error.message
+                }
+        if errors:
+            raise PreferenceValidationError(errors)
+        # Then perform the patch
+        for preference_key in update.keys():
+            preference_value = update[preference_key]
+            if preference_value is not None:
+                UserPreference.set_preference(existing_user, preference_key, preference_value)
+            else:
+                UserPreference.delete_preference(existing_user, preference_key)
+    except Exception as error:
+        raise PreferenceUpdateError(
+            "Error thrown when updating preferences: '{}'".format(error.message)
+        )
+
+
+def set_user_preference(requesting_user, preference_key, preference_value, username=None):
+    """Update a user preference for the given username.
+
+    Note:
+        It is up to the caller of this method to enforce the contract that this method is only called
+        with the user who made the request.
+
+    Arguments:
+        requesting_user (User): The user requesting to modify account information. Only the user with username
+            'username' has permissions to modify account information.
+        preference_key (string): The key for the user preference.
+        preference_value (string): The value to be stored.
+        username (string): Optional username specifying which account should be updated. If not specified,
+            `requesting_user.username` is assumed.
+
+    Raises:
+        AccountUserNotFound: no user with username `username` exists (or `requesting_user.username` if
+            `username` is not specified)
+        AccountNotAuthorized: the requesting_user does not have access to change the account
+            associated with `username`
+        AccountValidationError: the update was not attempted because validation errors were found with
+            the supplied update
+        PreferenceUpdateError: the update could not be completed.
+    """
+    existing_user = _get_user(requesting_user, username)
+    if preference_value is None or preference_value == '':
```

Comment:
What would an empty string mean for a preference? It seems to be equivalent to trying to remove the preference which should be handled using DELETE. Do you see a use case for an empty string having semantic meaning for a preference other than 'no preference'?